tests/conftest.py

```diff
@@ -28,3 +66,49 @@ async def client(app: FastAPI) -> AsyncIterator[AsyncClient]:
     """Return an ``httpx.AsyncClient`` configured to talk to the test app."""
     async with AsyncClient(app=app, base_url="https://example.com/") as client:
         yield client
+
+
+@pytest.fixture
+def k8s_storage_client(
+    obj_factory: TestObjectFactory,
+) -> MockK8sStorageClient:
+    return MockK8sStorageClient(test_obj=obj_factory)
+
+
+@pytest.fixture
+def docker_storage_client(
+    obj_factory: TestObjectFactory,
+) -> MockDockerStorageClient:
+    return MockDockerStorageClient(test_obj=obj_factory)
+
+
+@pytest_asyncio.fixture
+async def context_container(
+    config: Config,
+    client: AsyncClient,
+    obj_factory: TestObjectFactory,
+    k8s_storage_client: MockK8sStorageClient,
+    docker_storage_client: MockDockerStorageClient,
+) -> ContextContainer:
+    """Return a ``ContextContainer`` configured to supply dependencies."""
+    # Force K8s configuration to load
+    await initialize_kubernetes()
+    cc = ContextContainer.initialize(config=config, http_client=client)
+    # Patch container with storage mocks
+    cc.k8s_client = k8s_storage_client
+    cc.docker_client = docker_storage_client
```

Comment:
Ideally you'd mock the Kubernetes layer rather than mocking out the storage object, but sometimes it's more convenient to mock out the storage layer. But if you do that it should use dependency injection: the constructor of the service that uses the Kubernetes storage object should take the storage object to use, and you should have a mechanism for passing in the mock instead of the real version.

Reply:
Well, my initial plan, when I thought the storage layer was thinner than it's now going to end up, was just mock out the storage object, since it would basically be equivalent.  But if there's going to be more than backend-class-to-model translation going on there, yeah, probably mocking out the k8s/docker layer is a better idea.

Reply:
I think the approach I have isn't horrible, but it's going to need some tweaking.  Basically, the mocks are only going to be on the methods that actually use the underlying API.  This will work very well for Docker, which only lists tags and gets digests, and pretty well for K8s, where we have some object creation and deletion which doesn't really return much at the Python layer.  Making it work for the K8s event stream watchers will be the scary bit.